### PR TITLE
Inside out tutorial

### PR DESCRIPTION
**Pull request by manju956**
This tutorial defines step by step development of workload with file IO handling capability.
For the first input from the client, workload echoes user name along with encryption key.
For the subsequent requests, client submits user name along with key received in the previous step. As a response, client receives username along with number of hello world workload invocations by that user.
README file of hello_world tutorial is updated with steps to run the worlkload to access IO.

Also, implementation of DeleteFile and a wrapper for file IO is included as part of the PR.

Signed-off-by: manju956 <manjunath.a.c@intel.com>

### CODE REVIEW

File: docs/workload-tutorial/README.md [hunk 1]
```diff
@@ -253,3 +253,132 @@ To see what the updated source files should look like, refer to the files in
 directory
 [$TCF_HOME/docs/workload-tutorial/hello_world/stage_2/](hello_world/stage_2/).
 
+### <a name="phase3"></a>Phase 3: Worker-specific Code to execute IO opearations inside TEE
+In this example, we demonstrate IO execution by helloworld workload to store
+or retrive the count of workload hit by a user. The count value is stored in
```
Comment (danintel) [inline docs/workload-tutorial/README.md]: change retrive to retrieve

File: docs/workload-tutorial/README.md [hunk 2]
```diff
@@ -253,3 +253,132 @@ To see what the updated source files should look like, refer to the files in
 directory
 [$TCF_HOME/docs/workload-tutorial/hello_world/stage_2/](hello_world/stage_2/).
 
+### <a name="phase3"></a>Phase 3: Worker-specific Code to execute IO opearations inside TEE
+In this example, we demonstrate IO execution by helloworld workload to store
```
Comment (danintel) [inline docs/workload-tutorial/README.md]: change "by" to "by the"

File: docs/workload-tutorial/README.md [hunk 3]
```diff
@@ -253,3 +253,132 @@ To see what the updated source files should look like, refer to the files in
...
+or retrive the count of workload hit by a user. The count value is stored in
+the encrypted format while writing to file and decrypted while reading from file.
+
+For the first request, the client needs to submit the user name as in_data.  
+As a response, client receives `Hello <username> <encryption_key_in_hex>`.
+For the subsequent requests from the same user, client needs to submit username
+along with encryption key obtained in the previous step. As a response client receives,
+`Hello <username> [workload_hit_count]`.
+
+* Enter into the "Hello World" workload directory:
```
Comment (danintel) [inline docs/workload-tutorial/README.md]: Change "Enter into" to "Change to"

File: docs/workload-tutorial/README.md [hunk 4]
```diff
@@ -253,3 +253,132 @@ To see what the updated source files should look like, refer to the files in
 directory
 [$TCF_HOME/docs/workload-tutorial/hello_world/stage_2/](hello_world/stage_2/).
 
+### <a name="phase3"></a>Phase 3: Worker-specific Code to execute IO opearations inside TEE
```
Comment (danintel) [inline docs/workload-tutorial/README.md]: change "opearations" to "operations"
change "inside TEE" to "inside the TEE"

Comment (danintel) [reply docs/workload-tutorial/README.md]: I would add an introductory paragraph.  Your commit text with minor revisions is a good introduction and I would place it here. Something like:
```
This tutorial defines step by step development of a workload with file IO handling capability.
For the first input from the client, the workload echoes user name along with an encryption key.
For the subsequent requests, the client submits a user name along with a key received in the previous step.
As a response, the client receives a username along with the umber of hello world workload invocations by that user.

```

Comment (manju956) [reply docs/workload-tutorial/README.md]: I have updated the introductory paragraph at the beginning of phase 3.

File: docs/workload-tutorial/README.md (hunk 3 above)
Comment (manju956) [reply docs/workload-tutorial/README.md]: Done

File: docs/workload-tutorial/README.md (hunk 2 above)
Comment (manju956) [reply docs/workload-tutorial/README.md]: Done

File: docs/workload-tutorial/README.md (hunk 1 above)
Comment (manju956) [reply docs/workload-tutorial/README.md]: Done

**Comment by manju956** [comment]
> Thanks for adding this tutorial!
> 
> This PR also adds functionality:
> 
> * FileDelete
> * Read, Write, Delete with a filename parameter

I will edit my PR message to include the details.

File: tc/sgx/trusted_worker_manager/enclave_untrusted/enclave_bridge/file_io_processor.cpp [hunk 5]
```diff
@@ -406,3 +407,42 @@ uint32_t FileSeek(std::string fileName, size_t position, uint8_t *result,
...
+
+    try {
+        if ( remove(fileName.c_str()) != 0 ) {
+            tcf::Log(TCF_LOG_DEBUG, "File %s delete failed", fileName.c_str());
+            resultStr = "FILE DELETE FAILED";
+            resultStrSize = resultStr.length();
+            result = (uint8_t *)strncpy((char *)result, resultStr.c_str(), resultStrSize);
+            tcf::error::ThrowIf<tcf::error::ValueError>(
+                resultStrSize > resultSize, "result string size is "
+                "greater than max result size");
```
Comment (manojgop) [inline tc/sgx/trusted_worker_manager/enclave_untrusted/enclave_bridge/file_io_processor.cpp]: We should check for error condition before strncpy and throw error

File: common/sgx_workload/iohandler/file_io_wrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    uint32_t file_handler_id = file_io.GetIoHandlerId("tcf-base-file-io");
+    file_io.SetIoHandlerId(file_handler_id);
+    file_io.SetFileName(file_name);
+    std::string result;
+    size_t result_size = file_io.GetMaxIoResultSize();
+    result.reserve(result_size);
+
+    uint32_t status = file_io.FileDelete((uint8_t *)result.c_str(), result_size);
+    return status;
+}
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx_workload/iohandler/file_io_wrapper.cpp]: Fixed

File: common/sgx_workload/iohandler/file_io_wrapper.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+* limitations under the License.
+*/
+
+#include <string>
+
+std::string Read(std::string file_name);
+
+uint32_t Write(std::string file_name, std::string data);
+
+uint32_t Delete(std::string file_name);
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx_workload/iohandler/file_io_wrapper.h]: Fixed

File: common/sgx_workload/iohandler/file_io_wrapper.h [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include <string>
+
+std::string Read(std::string file_name);
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx_workload/iohandler/file_io_wrapper.h]: Done

File: tc/sgx/trusted_worker_manager/enclave_untrusted/enclave_bridge/file_io_processor.cpp (hunk 5 above)
Comment (manju956) [reply tc/sgx/trusted_worker_manager/enclave_untrusted/enclave_bridge/file_io_processor.cpp]: Done

**Pull request merged**